packages/connection/src/class-nonce-handler.php

```diff
@@ -0,0 +1,197 @@
+<?php
+/**
+ * The nonce handler.
+ *
+ * @package automattic/jetpack-connection
+ */
+
+namespace Automattic\Jetpack\Connection;
+
+/**
+ * The nonce handler.
+ */
+class Nonce_Handler {
+
+	/**
+	 * How many nonces should be removed during each run of the runtime cleanup.
+	 * Can be modified using the filter `jetpack_connection_nonce_cleanup_runtime_limit`.
+	 */
+	const CLEANUP_RUNTIME_LIMIT = 10;
+
+	/**
+	 * How many nonces should be removed per batch during the `clean_all()` run.
+	 */
+	const CLEAN_ALL_LIMIT_PER_BATCH = 1000;
+
+	/**
+	 * Nonce lifetime in seconds.
+	 */
+	const LIFETIME = HOUR_IN_SECONDS;
+
+	/**
+	 * The nonces used during the request are stored here to keep them valid.
+	 *
+	 * @var array
+	 */
+	private static $nonces_used_this_request = array();
+
+	/**
+	 * Adds a used nonce to a list of known nonces.
+	 *
+	 * @param int    $timestamp the current request timestamp.
+	 * @param string $nonce the nonce value.
+	 * @param bool   $run_cleanup Whether to run the `cleanup_runtime()`.
+	 *
+	 * @return bool whether the nonce is unique or not.
+	 */
+	public static function add( $timestamp, $nonce, $run_cleanup = true ) {
+		global $wpdb;
+
+		if ( isset( static::$nonces_used_this_request[ "$timestamp:$nonce" ] ) ) {
+			return static::$nonces_used_this_request[ "$timestamp:$nonce" ];
+		}
+
+		// This should always have gone through Jetpack_Signature::sign_request() first to check $timestamp an $nonce.
+		$timestamp = (int) $timestamp;
+		$nonce     = esc_sql( $nonce );
+
+		// Raw query so we can avoid races: add_option will also update.
+		$show_errors = $wpdb->show_errors( false );
+
+		$old_nonce = $wpdb->get_row(
+			$wpdb->prepare( "SELECT * FROM `$wpdb->options` WHERE option_name = %s", "jetpack_nonce_{$timestamp}_{$nonce}" )
+		);
+
+		if ( is_null( $old_nonce ) ) {
+			$return = (bool) $wpdb->query(
+				$wpdb->prepare(
+					"INSERT INTO `$wpdb->options` (`option_name`, `option_value`, `autoload`) VALUES (%s, %s, %s)",
+					"jetpack_nonce_{$timestamp}_{$nonce}",
+					time(),
+					'no'
+				)
+			);
+
+			/**
+			 * Use the filter to disable the nonce cleanup that happens at shutdown after adding a new nonce.
+			 *
+			 * @since 9.0.0
+			 *
+			 * @param int $limit How many old nonces to remove at shutdown.
+			 */
+			if ( apply_filters( 'jetpack_connection_add_nonce_cleanup', $run_cleanup ) ) {
+				add_action( 'shutdown', array( __CLASS__, 'clean_runtime' ) );
+			}
+		} else {
+			$return = false;
+		}
+
+		$wpdb->show_errors( $show_errors );
+
+		static::$nonces_used_this_request[ "$timestamp:$nonce" ] = $return;
+
+		return $return;
+	}
+
+	/**
+	 * Removing all the nonces.
+	 *
+	 * @param int $cutoff_timestamp All nonces added before this timestamp will be removed.
+	 *
+	 * @return true
+	 */
+	public static function clean_all( $cutoff_timestamp = PHP_INT_MAX ) {
+		for ( $i = 0; $i < 1000; ++$i ) {
```

Comment:
Why 1000?

I guess the idea is to not run for a very long time if there are millions of nonces, but it does seem to belie the stated contract that this deletes *all* nonces.

Reply:
I think it's unrealistic to expect over a million of nonces to be removed at one request.
That way the person will never be able to disconnect Jetpack.

Good catch though, I'll add a comment mentioning that limit.

Reply:
I just ran into a disconnect request timeout because of this function. The `clean_all` run would have probably caused the same problem on an old code. Since it's the last chance to remove everything, could we rely on the index and instead do something like:
```
DELETE FROM wp_options WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < 'jetpack_nonce_ZZZZZZZZZZZZZZZZZZZZZZZZZZZZ'
```

Reply:
Just ran some tests, and it doesn't look good for sites with more than about 500k records:
```
MariaDB [wordpress_70]> DELETE FROM wp_options WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < 'jetpack_nonce_ZZZZZZZZZZZZZZZZZZZZZZZZZZZZ';
Query OK, 5000 rows affected (0.234 sec)

MariaDB [wordpress_70]> DELETE FROM wp_options WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < 'jetpack_nonce_ZZZZZZZZZZZZZZZZZZZZZZZZZZZZ';
Query OK, 1790000 rows affected (11 min 22.798 sec)
```

Reply:
Right, what I'm mostly concerned about is the timeout when disconnecting. We try to clean up everything as much as possible, but in case of a site with many nonces stored, this may cause a timeout. We don't really have a good option here as far as I can see. We're going to leave the site with leftover notices anyway, but we can at least try to clean up as much as we can. The million entries that the method is limited to right now in case of a disconnect is not going to work, I think we need some sort of time constraint here to avoid timeouts. What do you think @sergeymitr ?

Reply:
@zinigor
I like the idea.
Since default maximum execution time is 30 seconds, we could limit it to 20 seconds in addition to existing limit (1k iterations 1k records each).

So if not all records are removed in 20 seconds, we can't really do much with it. We could add wp-cron cleanup job specifically for this use case, but I don't think it's worth the added complexity. If somebody disconnects Jetpack, they're most likely going to disable or remove it anyway.

Something like this:
```suggestion
		for ( $time = time(), $i = 0; $i < 1000 && time() > $time - 20; ++$i ) {
			// ...
		}
```

Reply:
I added the time limit: https://github.com/Automattic/jetpack/pull/17030/files#diff-c856b8c22febf99e6f16144e03d1d277R109